### PR TITLE
[Feature] Add aux and last hidden state dumping for EAGLE

### PR DESCRIPTION
**Pull request by ChangseokSong**
## Motivation

This PR adds a feature to dump auxiliary and last hidden states produced during EAGLE speculative decoding.  
The main purpose is to support SpecForge offline training, and it can also be used to collect data from real requests for training draft models.  

---

### Hidden-state capture
- Auxiliary and last hidden states are captured in the logits processor.
- Hidden states for each request are moved to host once, after the request finishes.

### Dumping workflow
- Added `HiddenStateDumper`, which manages preparing payloads, staging tensors on CPU, and writing them to disk.
- Disk writes are handled by a process pool to avoid blocking the main worker.
- A separate CUDA stream is used for DtoH transfers so that they can overlap with verification.
- Hidden states from the previous decode step are processed at the next step, and the final batch is flushed when all requests in the batch are finished.
- Dumping is performed in a round robin manner across TP ranks.

### Buffer reuse
- Added `FlatBufferPool`, which reuses CPU buffers to reduce allocation overhead.
 

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/speculative/eagle_utils.py [hunk 1]
```diff
@@ -197,3 +203,241 @@ def verify_tree_greedy_func(
...
+            return
+
+        assert (
+            self.server_args.speculative_eagle_hidden_states_dump_path is not None
+        ), "speculative_eagle_hidden_states_dump_path must be set"
+
+        if self.server_args.speculative_eagle_dump_accept_rate_threshold < 1.0:
+            acceptance_rate = (req.spec_accepted_tokens + req.spec_verify_ct) / (
+                req.spec_verify_ct * self.server_args.speculative_num_draft_tokens
+            )
```
[the remark this answers is not in the document]
Comment (ChangseokSong) [reply python/sglang/srt/speculative/eagle_utils.py]: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/managers/tokenizer_manager.py#L1844
Calculate acceptance rate: accepted / (steps * lookahead)

File: python/sglang/srt/speculative/eagle_utils.py [hunk 2]
```diff
@@ -197,3 +203,241 @@ def verify_tree_greedy_func(
...
+        )
+
+
+def dump_hidden_states(
+    dump_path: str,
+    hidden_states: torch.Tensor,
+    aux_hidden_states: torch.Tensor,
+    origin_input_ids: List[int],
+    output_ids: List[int],
+):
```
[the remark this answers is not in the document]
Comment (ChangseokSong) [reply python/sglang/srt/speculative/eagle_utils.py:372]: fixed

File: python/sglang/srt/speculative/eagle_worker.py [hunk 3]
```diff
@@ -676,6 +687,12 @@ def verify(self, batch: ScheduleBatch, spec_info: EagleVerifyInput):
             batch_result.can_run_cuda_graph,
         )
 
+        if (
+            self.server_args.speculative_eagle_enable_dump_hidden_states
+            and self.dump_worker.payloads is not None
+        ):
+            self.dump_worker.process_dump_payload()
```
[the remark this answers is not in the document]
Comment (ChangseokSong) [reply python/sglang/srt/speculative/eagle_worker.py:710]: It's placed there so the CPU work from `process_dump_payload()` can hide under the GPU compute from `self.target_worker.forward_batch_generation`.

**Comment by hnyls2002** [comment]
Hi @ChangseokSong, this PR targets the Spec V1 eagle_worker.py, which was removed from main when Spec V1 was deprecated in #25464, so the implementation no longer applies. Closing as obsolete; a hidden-state dumping feature for the V2 worker would need a fresh PR. Please reopen if I've missed something.
